test/unit/web.c

```diff
@@ -0,0 +1,227 @@
+    int safe = ctx_is_safe[ctxind];
+    if(!safe) {
+      pthread_mutex_lock(&ctx_locks[ctxind]);
+    }
+    shmem_ctx_long_p(contexts[ctxind],&c->can_send,1L,prevpe);
+    if(!safe) {
+      pthread_mutex_unlock(&ctx_locks[ctxind]);
+    }
+
+    shmem_wait_until(&c->can_send,SHMEM_CMP_NE,0L);
+    c->can_send = 0;
+
+    ctxind = ((unsigned)rand_r(&c->seed))%NUM_CONTEXTS;
+    safe = ctx_is_safe[ctxind];
+
+    if(!safe) {
+      pthread_mutex_lock(&ctx_locks[ctxind]);
+    }
+    shmem_ctx_long_p(contexts[ctxind],&c->rec_val,c->send_val,nextpe);
+    shmem_ctx_fence(contexts[ctxind]);
+    shmem_ctx_long_p(contexts[ctxind],&c->received,1L,nextpe);
+    if(!safe) {
+      pthread_mutex_unlock(&ctx_locks[ctxind]);
+    }
+
+    shmem_wait_until(&c->received,SHMEM_CMP_NE,0L);
+    c->received = 0;
+    c->send_val = c->rec_val;
+  }
+
+  /* printf("[%d:%d] done\n",me,c->idx); */
+
+  return 0;
+}
+
+int main(int argc, char* argv[]) {
+    int i;
+    for(i = 0; i < NUM_CONTEXTS; ++i) {
+      int err;
+      err = pthread_mutex_init(&ctx_locks[i],NULL);
+      if (err) {
+          perror("Mutex initialization failed");
+          return 1;
+      }
+    }
+
+    int err;
+    int errors = 0;
+    int tl = SHMEM_THREAD_MULTIPLE, tl_provided;
+
+    shmem_init_thread(tl,&tl_provided);
+    CHECK_ASSERT(tl == tl_provided);
+
+    n_pes = shmem_n_pes();
+    me = shmem_my_pe();
+
+    if (n_pes < 2) {
+        if (me == 0)
+            printf("web: Requires 2 or more PEs\n");
+        shmem_finalize();
+        return 0;
+    }
+
+    srand(1+me);
+
+    n_threads = 4 * n_pes;
+    channels = shmem_malloc(n_threads*sizeof(channel));
+    memset(channels,0,n_threads*sizeof(channel));
+
+    long* vals = malloc(sizeof(long)*n_threads);
+    for(i = 0; i < n_threads; ++i) {
+      channels[i].idx = i;
+      channels[i].received = 0;
+      channels[i].can_send = 0;
+      channels[i].rec_val = 0;
+      channels[i].send_val = vals[i] = rand();
+
+      channels[i].stride = (i)%(n_pes-1) + 1;
+      channels[i].seed = 1 + rand() + i;
+    }
+
+    max_steps = get_max_steps(n_pes);
+
+    for(i = 0; i < NUM_CONTEXTS; ++i) {
+      ctx_is_safe[i] = (i >= NUM_UNSAFE_CTX);
+      err = shmem_ctx_create(ctx_is_safe[i] ? 0 : SHMEM_CTX_SERIALIZED,&contexts[i]);
+      CHECK_ERROR(err,"Failed to create ctx %d\n",i);
+    }
+
+    shmem_barrier_all();
+
+    pthread_t* threads = calloc(sizeof(pthread_t),n_threads);
+    for(i = 0; i < n_threads; ++i) {
+      int err;
+      err = pthread_create(&threads[i],NULL,runchannel,&channels[i]);
+      CHECK_ASSERT(0 == err);
+    }
+
+    for (i = 0; i < n_threads; i++) {
+        int err;
+        err = pthread_join(threads[i], NULL);
+        CHECK_ASSERT(0 == err);
+    }
+    free(threads);
+
+    for(i = 0; i < NUM_CONTEXTS; ++i) {
+      shmem_ctx_quiet(contexts[i]);
+      shmem_ctx_destroy(contexts[i]);
+      int err;
+      err = pthread_mutex_destroy(&ctx_locks[i]);
+      CHECK_ASSERT(!err);
+    }
+
```

Comment:
Do you need a `shmem_sync_all` here?  If not, I don't quite understand why it's needed at the end of  `many-ctx.c` and not here...

Reply:
Yup, good catch!

Reply:
Scratch that.  Web takes care of this via the ```received``` that is waited on in web.c:124.